### PR TITLE
Do a small builder refactor.

### PR DESCRIPTION
**Pull request by ericholscher**
This adds:

* A Local environment for checkout & config reading (Host)
* A Build environment for pip/conda & doc building (Docker)
* Refactors how we pass Bash environments around 
* Cleans up the python environment to use internal state

### CODE REVIEW

File: readthedocs/doc_builder/python_environments.py [hunk 1]
```diff
@@ -180,7 +180,6 @@ def setup_base(self):
             '--name',
             self.version.slug,
             'python={python_version}'.format(python_version=self.config.python_version),
```
Comment (jakirkham) [inline readthedocs/doc_builder/python_environments.py]: Probably can lose this comma, but its not a big deal.

**Comment by jakirkham** [comment]
Looks cool. Should there be something for virtualenv, as well? Not sure if that's being used. It seems to just add the correct python to the path so would be simple. ( https://github.com/pypa/virtualenv/blob/18b374dc3eeb0e37aa3ad1b22c442712d9dca103/virtualenv_embedded/activate.sh )

**Comment by ericholscher** [comment]
True. We're still injecting the bin_path manually, because it varies by command, and we don't want to override the default $PATH -- so bin path's will still be there. We should be setting the Virtualenv default PATH though, +1

File: readthedocs/projects/tasks.py [hunk 2]
```diff
@@ -229,6 +222,43 @@ def setup_vcs(self):
             raise BuildEnvironmentError('Failed to import project',
                                         status_code=404)
 
+    def get_bash_env(self):
+        """
+        Get bash environment variables used for all builder commands.
```
Comment (agjohnson) [inline readthedocs/projects/tasks.py]: `BuildCommand` defaults to `shell=False`, which should be what we're using for almost all commands. Naming/referencing `bash` here is misleading, these are just env variables as far as subprocess is concerned.

File: readthedocs/projects/tasks.py [hunk 3]
```diff
@@ -126,12 +125,20 @@ def run(self, pk, version_pk=None, build_pk=None, record=True, docker=False,
                     status_code=423
                 )
 
+            self.config = load_yaml_config(version=self.version)
+
+        bash_env = self.get_bash_env()
```
Comment (agjohnson) [inline readthedocs/projects/tasks.py]: Same note regarding `bash` here

**Comment by agjohnson** [comment]
It might be good to include a test to ensure how we're settings `PATH` now. Besides my other notes, this looks good though.

File: readthedocs/projects/tasks.py (hunk 2 above)
Comment (ericholscher) [inline readthedocs/projects/tasks.py]: We need some way to name them different than the 2 other types of environments we use. Any suggestions?

Comment (agjohnson) [inline readthedocs/projects/tasks.py]: `env_vars` + `get_environment_variables` would make sense here

**Comment by jakirkham** [comment]
The only other point I would raise is the same one we already discussed, @ericholscher. That is we now may have an environment active before it is created. I'm not sure what this will do for virtualenv. With `conda`, I don't think it will be an issue as long as all `conda` commands before the environment  explicitly exist use the option `-n` to specify the environment name. Though it would be good to check just to be sure.

**Comment by ericholscher** [comment]
@jakirkham I went ahead and fixed this by passing an explicit bin_path=None to the environment creation commands. Nice catch.

**Pull request merged**